WT-13076 Repurpose the time aggregate values to track page level stop timestamp

kommiharibabu opened the pull request:
Rename the existing newest_stop_durable_ts into newest_page_stop_durable_ts to track the durable stop timestamp of a page when the entire page is removed instead of tracking the maximum stop durable timestamps on a page.

Rename the existing newest_start_durable_ts into newest_durable_ts to track both start and stop durable timestamps.

kommiharibabu committed:
WT-13076 Repurpose the time aggregate values to track page level stop timestamp

Rename the existing newest_stop_durable_ts into newest_page_stop_durable_ts
to track the durable stop timestamp of a page when the entire page is removed
instead of tracking the maximum stop durable timestamps on a page.

Rename the existing newest_start_durable_ts into newest_durable_ts to include
both start and stop durable timestamp.

luke-pearson commented:
@kommiharibabu I was under the impression that aggregated timestamps already apply to a whole page. Which leads me to the question of what is the difference between the new "newest_page_stop_durable_ts" and the existing "newest_stop_durable_ts" which already aggregated timestamps on a page?

kommiharibabu commented:
@luke-pearson Yes, the aggregated timestamps are based on the whole page. Suppose when the page has half of the keys removed, the `newest_stop_durable_ts` contains the maximum durable timestamp from those tombstones, but the `newest_page_stop_durable_ts` will have only a maximum durable timestamp when the whole page is removed.

The `newest_stop_durable_ts` is mainly used by the checkpoint cleanup to remove obsolete pages. The checkpoint cleanup can remove a page when every key in the page is removed. But this `newest_stop_durable_ts` doesn't provide clear information on whether a checkpoint cleanup can succeed or not. Suppose one key is removed on a page and it is not removed from the page and a stop is written to the disk, the checkpoint cleanup keeps on visiting that page, which is not required.

quchenhao commented on src/btree/bt_sync_obsolete.c:
```diff
@@ -475,7 +476,7 @@ __checkpoint_cleanup_eligibility(WT_SESSION_IMPL *session, const char *uri, cons
     __wt_config_subinit(session, &ckptconf, &cval);
     while ((ret = __wt_config_next(&ckptconf, &key, &cval)) == 0) {
-        ret = __wt_config_subgets(session, &cval, "newest_stop_durable_ts", &value);
+        ret = __wt_config_subgets(session, &cval, "newest_page_stop_durable_ts", &value);
```
How it is backward compatible if this is changed in checkpoint?

kommiharibabu replied:
There is no problem with backward compatibility, here we just ignore this file until the next checkpoint occurs. When the checkpoint happens, all the data is written in the newer format and it works fine.

The only problem with the current code is downgrading to an older version using an unclean shutdown.

quchenhao replied:
> The only problem with the current code is downgrading to an older version using an unclean shutdown.

Yes, this is what I'm worried about. We cannot introduce this to 8.0 as we have a plan to be able to directly upgrade from 5.0 to 8.0.

kommiharibabu replied:
How about writing the old and new format both for one major version and removing it later? It an extra data for the checkpoint, other than it is not going to cause any problem.

quchenhao replied:
At the checkpoint level, writing both I think solves the issue. But still do we know will it still wort that if the database have pages with timestamps of different meanings? Also we don't know which page has what types of timestamps.

kommiharibabu replied:
Writing both variable names for the checkpoint has solved the upgrade/downgrade problem. 

> But still do we know will it work that if the database have pages with timestamps of different meanings? Also we don't know which page has what types of timestamps.

We are not changing how we store the timestamp information on the page/cell. It is jut the value that is changed, so there is no problem.

agorrod reviewed (commented):
I've asked some questions in the Jira ticket.

kommiharibabu committed:
Fix the assert

kommiharibabu committed:
set txn value also before checking it

kommiharibabu committed:
Merge branch 'develop' into wt-13076-repurpose-time-aggregate

kommiharibabu committed:
Merge branch 'develop' into wt-13076-repurpose-time-aggregate

kommiharibabu committed:
macro rename miss

kommiharibabu committed:
Merge branch 'develop' into wt-13076-repurpose-time-aggregate

kommiharibabu committed:
Fix a couple of issues

kommiharibabu committed:
Merge branch 'develop' into wt-13076-repurpose-time-aggregate

kommiharibabu committed:
Merge branch 'develop' into wt-13076-repurpose-time-aggregate

kommiharibabu committed:
fix s_all

kommiharibabu committed:
Merge branch 'develop' into wt-13076-repurpose-time-aggregate

kommiharibabu committed:
WT-13076 Fix incorrect time aggregate validation and style issues

Remove the newly-added assertion that newest_stop_ts <= newest_durable_ts.
This assertion fires for prepared deletes because newest_durable_ts is only
updated for committed stop transactions, while newest_stop_ts is set to
stop_prepare_ts for prepared transactions. The existing check against
newest_page_stop_durable_ts already covers the committed case correctly.

Also fix a non-ASCII em dash in a comment and consolidate fragmented string
literals in validation error messages.

kommiharibabu committed:
WT-13076 Update newest_durable_ts for prepared stop transactions

In WT_TIME_AGGREGATE_UPDATE, update newest_durable_ts with stop_prepare_ts
in the prepared-stop branch, mirroring what the original newest_stop_durable_ts
did. This maintains the invariant newest_durable_ts >= newest_stop_ts for both
committed and prepared transactions, allowing the validation assertion to fire
correctly for bugs while not triggering for legitimate prepared deletes.

kommiharibabu committed:
WT-13076 Fix checkpoint cleanup eligibility on insert-only btrees

The metadata writer keeps the legacy newest_stop_durable_ts field for
backward compatibility but populates it from newest_durable_ts as a safe
over-estimate. The eligibility reader was merging that legacy field into
newest_page_stop_durable_ts, which has stricter semantics (set only for
full-page deletes). An insert-only btree therefore looked like it had
deletes, eligibility returned true, and __sync_obsolete_tw_check
triggered disk reads for pages whose newest_durable_ts had fallen below
oldest_ts. Merge the legacy field into newest_durable_ts instead, which
matches how the metadata writer derives it.

kommiharibabu committed:
Merge remote-tracking branch 'origin/develop' into wt-13076-repurpose-time-aggregate

# Conflicts:
#	src/include/wiredtiger.h.in

kommiharibabu committed:
Merge remote-tracking branch 'origin/develop' into wt-13076-repurpose-time-aggregate

# Conflicts:
#	src/include/cell_inline.h
#	src/include/wiredtiger.h.in
#	src/meta/meta_ckpt.c

kommiharibabu committed:
Merge develop into wt-13076-repurpose-time-aggregate

kommiharibabu committed:
Add checkpoint cleanup benchmark design

kommiharibabu committed:
Clarify cold-cache benchmark setup

kommiharibabu committed:
Plan checkpoint cleanup cache benchmark

kommiharibabu committed:
Fix page stop timestamp in range visibility check.

kommiharibabu committed:
Add checkpoint cleanup cache pollution benchmark

kommiharibabu committed:
Add checkpoint cleanup timestamp test

kommiharibabu committed:
Revert "Clarify cold-cache benchmark setup"

This reverts commit 4146026cc3f6e7deb43f31ac5a66255092e608e4.

kommiharibabu committed:
Revert "Plan checkpoint cleanup cache benchmark"

This reverts commit de094cda2c269dc0735ebc7538438216aa014ba8.

kommiharibabu committed:
Revert "Add checkpoint cleanup cache pollution benchmark"

This reverts commit bf3b53f49a92c79f08aabefcafb6758872bd13f0.

kommiharibabu committed:
Revert "Add checkpoint cleanup benchmark design"

This reverts commit 19c5f094276539b46dd902d755a51913e6cce7cb.

kommiharibabu committed:
Clarify checkpoint cleanup page skip policy

kommiharibabu committed:
Fix fast style validation issues

kommiharibabu committed:
Guard page stop timestamp validation

kommiharibabu committed:
Restore legacy page stop timestamp semantics

kommiharibabu committed:
Skip checkpoint cleanup test for disagg hook

kommiharibabu committed:
Revert "Restore legacy page stop timestamp semantics"

This reverts commit eb90d02bc810029c0b568ac374515ee7281bb519.

kommiharibabu committed:
Include page stop timestamp in checkpoint aggregate

kommiharibabu committed:
Format checkpoint aggregate load

kommiharibabu committed:
Clarify legacy checkpoint metadata comment

kommiharibabu committed:
Preserve durable timestamp aggregate on cell load

kommiharibabu committed:
Normalize page stop aggregate on cell load

kommiharibabu committed:
Remove page stop reset during cell load

kommiharibabu committed:
WT-13076 extend compatibility coverage

kommiharibabu committed:
WT-13076 verify files across compatibility cycles

kommiharibabu committed:
WT-13076 preserve aggregate compatibility

kommiharibabu committed:
WT-13076 include 9.0 in compatibility suite

kommiharibabu committed:
WT-13076 preserve parent aggregate timestamps

kommiharibabu committed:
WT-13076 fix compatibility test round sequencing
